lib/rubygems/commands/build_command.rb

```diff
@@ -45,7 +49,7 @@ def execute
       spec = Gem::Specification.load gemspec
 
       if spec then
-        Gem::Package.build spec, options[:force]
+        Gem::Package.build spec, options[:force], options[:allow_warnings]
```

Comment:
You should be able to use [this test](https://github.com/rubygems/rubygems/blob/master/test/rubygems/test_gem_commands_build_command.rb#L27-L51) as a base for testing `--allow-warnings`